### ISSUE TITLE
OADP does not allow install Velero with feature flag

### ISSUE THREAD

**Issue by phuongatemc** [konveyor/oadp-operator#66]
We want to use velero-plugin-for-vpshere https://github.com/vmware-tanzu/velero-plugin-for-vsphere but it requires additional feature flag (--features=EnableVSphereItemActionPlugin) to function properly.  However, OADP won't allow setting any feature flag on the Web-UI when install Velero.  This prevents us from using the vSphere plugin mentioned above.

**Comment by dymurray** [comment]
@phuongatemc FYI we dug intoo this and this seemed to be the case for an older version of the vsphere plugin. You should be able to use OADP 0.2.0 with the latest plugin and not need to set this feature flag.

please let me know if you run into any other issues.

**Comment by dymurray** [comment]
Also FYI, I don't believe the vsphere plugin supports openshift yet. I could be wrong.

**Comment by seema-emc** [comment]
@dymurray With the latest vSphere plugin, we need the following feature flag to make sure Vsphere plugin's snapshot manager comes in LocalMode.
        - --features=EnableLocalMode

**Comment by shubham-pampattiwar** [comment]
@seema-emc @phuongatemc I have OADP operator changes ready with me that would enable you to set the value for the `--features` flag. I tried out the following things for enabling the velero plugin for vsphere via the OADP operator:
- Set the `--features=EnableLocalMode` I believe this is consumed by the velero plugin for vsphere code [here](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/3f5e361f4f46394c2055c443eb9a368e042b8470/pkg/cmd/datamgr/cli/install/install.go#L154) 
- Set up an init container of velero plugin for vsphere using the [image](https://hub.docker.com/r/vsphereveleroplugin/velero-plugin-for-vsphere/tags?page=1&ordering=last_updated) 
- Used upstream Velero [image](https://hub.docker.com/r/velero/velero/tags?page=1&ordering=last_updated) for Velero deployment

After doing the above steps I was able to get the following logs in the Velero pod created by the OADP operator (implying that the velero plugin for vsphere is getting registred):
```
shubham@localhost ~/oadp-operator -  (master) $ oc logs velero-6cfc5f7f7c-77vsq | grep vsphere
time="2021-03-11T07:50:29Z" level=info msg="registering plugin" command=/plugins/velero-plugin-for-vsphere kind=BackupItemAction logSource="pkg/plugin/clientmgmt/registry.go:100" name=velero.io/vsphere-pvc-backupper
time="2021-03-11T07:50:29Z" level=info msg="registering plugin" command=/plugins/velero-plugin-for-vsphere kind=VolumeSnapshotter logSource="pkg/plugin/clientmgmt/registry.go:100" name=velero.io/vsphere
time="2021-03-11T07:50:29Z" level=info msg="registering plugin" command=/plugins/velero-plugin-for-vsphere kind=RestoreItemAction logSource="pkg/plugin/clientmgmt/registry.go:100" name=velero.io/vsphere-pvc-restorer
time="2021-03-11T07:50:29Z" level=info msg="registering plugin" command=/plugins/velero-plugin-for-vsphere kind=DeleteItemAction logSource="pkg/plugin/clientmgmt/registry.go:100" name=velero.io/vsphere-pvc-deleter

```
Do the above solution steps and log output fulfill your use case ? Do you just want the ability to set the `--features` flag values to Velero Deployment Pod or you also want the velero plugin for sphere initContainer creation as well alongside the ability to set features flag ? 
Let me know if this solution suffices your use case or If I am misinterpreting or missing anything.

**Comment by phuongatemc** [comment]
@shubham-pampattiwar  Do I need a new/private OADP operator binary from somewhere or simply delete the OADP operator and the latest one with this fix will be used?

**Comment by shubham-pampattiwar** [comment]
@phuongatemc Once this fix is merged into master there are two ways you can use the latest operator:
1. [Delete your existing operator](https://github.com/konveyor/oadp-operator#cleanup) and redeploy using the [Non-OLM steps](https://github.com/konveyor/oadp-operator#getting-started-with-basic-install-non---olm) 
OR
2.  As We will be doing a minor release of the operator (i.e. once the latest changes are merged in the master we update the operator in community operators), you can use the latest operator via OperatorHub present on OpenShift UI Console after cleaning up the older one.

Also, @phuongatemc Can you please confirm if the proposed solution is the one you wanted and satisfies your usecase ?

**Comment by phuongatemc** [comment]
The approach looks good to me.  Thank you!

**Comment by shubham-pampattiwar** [comment]
@phuongatemc Can you make sure that you are using OADP operator version `v0.2.1` ? The operator pod's container image tag should be `v0.2.1`

**Comment by shubham-pampattiwar** [comment]
Also, @phuongatemc When using the parameter `velero_feature_flags` , please keep in mind that it is not a list of flags. So instead of using it as :
```
velero_feature_flags:
- EnableLocalMode
```
Use it as follows (correct usage, treat it is as comma separated string for multiple values):
```
velero_feature_flags: EnableLocalMode
```
Reference docs: https://github.com/konveyor/oadp-operator#usage-of-velero---features-option

**Comment by phuongatemc** [comment]
> @phuongatemc Can you make sure that you are using OADP operator version `v0.2.1` ? The operator pod's container image tag should be `v0.2.1`

Nope.  I have v0.2.0.  I just ran "Install" on a new OpenShift Cluster and it pulled in that version.  Is there anyway I can specify it to pull v0.2.1?

**Comment by shubham-pampattiwar** [comment]
@phuongatemc you should be able to get the latest version from the OperatorHub itself. There are 2 things that you can try out:
1. Cleanup the existing instance of the operator using the steps specified here https://github.com/konveyor/oadp-operator#cleanup and try to install it again from the OperatorHub from OpenShift console, the operatorHub must have the latest version as `0.2.1`
2. Or You can try out the Alternative install approach by following Non-OLM installation steps specified here: https://github.com/konveyor/oadp-operator#getting-started-with-basic-install-non---olm

**Comment by jwmatthews** [comment]
@phuongatemc check what shows up from OperatorHub, maybe you need to refresh the Catalog or something.

Below is an example of what I see, showing v0.2.1 being available.
<img width="1577" alt="Screen Shot 2021-03-31 at 6 40 14 PM" src="https://user-images.githubusercontent.com/206875/113220118-a2205900-9250-11eb-93d5-a32fdd254587.png">

**Comment by phuongatemc** [comment]
![image](https://user-images.githubusercontent.com/24380334/113224246-df301f80-923f-11eb-897d-e55691d1a2f9.png)
Look like my Catalog is out-of-date because it contain v0.2.0.  How can I update my Catalog?

**Comment by phuongatemc** [comment]
Quick update:
1. Try to clean up the oadp-operator namespace and retry install OADP, I still got the version v0.2.0.  
2. Try to manually download the OADP source code and build the OADP ==> failed with below error:
 podman build -f build/Dockerfile . -t oadp-operator:latest
STEP 1: FROM quay.io/operator-framework/ansible-operator:v0.17.0
Getting image source signatures
Copying blob 9d05b535e611 done
Copying blob a7518fc4b4b8 done
Copying blob 1730fea3a412 done
Copying blob 51a84b8e6e04 done
Copying blob ee2244abc66f done
Copying blob befb03b11956 done
Copying blob 96917d26f5e1 done
Copying blob 45b98e5f36b9 done
Copying blob a5236a9afd95 done
Copying config 5d34c0760f done
Writing manifest to image destination
Storing signatures
STEP 2: COPY requirements.yml ${HOME}/requirements.yml
86d2deb5aaa8d76f16c50a82607f981a8320988c91596368b5191db3fa06c96e
STEP 3: RUN ansible-galaxy collection install -r ${HOME}/requirements.yml  && chmod -R ug+rwx ${HOME}/.ansible
Process install dependency map
**ERROR! Unknown error when attempting to call Galaxy at 'https://galaxy.ansible.com/api/': <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:897)>
Error: error building at STEP "RUN ansible-galaxy collection install -r ${HOME}/requirements.yml  && chmod -R ug+rwx ${HOME}/.ansible": error while running runtime: exit status 1**

**Comment by pkchavva** [comment]
Team I am checking with the Operator team to understand and debug this further. Will keep you posted.

**Comment by shubham-pampattiwar** [comment]
@phuongatemc you do not need to build the image, building image and using that operator image is only needed if you changing the operator's code and want to test the same. For your use case you can directly install the operator by following the installation steps specified here https://github.com/konveyor/oadp-operator#operator-installation
Please let me know if these steps do not work out for you.

**Comment by phuongatemc** [comment]
@shubham-pampattiwar:  Run operation without build works https://github.com/konveyor/oadp-operator#operator-installation .  Thank you.

**Comment by phuongatemc** [comment]
@shubham-pampattiwar 
The GUI doesn't recognize the OADP installed manually as mentioned above.  As consequence, I cannot run backup via this OADP:
![image](https://user-images.githubusercontent.com/24380334/113360298-53c99380-92fe-11eb-985e-f145101aca4c.png)

**Comment by shubham-pampattiwar** [comment]
@phuongatemc Yes that might be the case as the operator is not installed via OLM (OperatorHub), therefore Operator UI would not be available. But you can always create a backup resource using a yaml file via command line (`oc create -f  <backup-yaml-files>`). For references you can see the examples here https://github.com/konveyor/velero-examples, Please let me know if this works for you.

**Comment by phuongatemc** [comment]
@shubham-pampattiwar 
I found out that my environment missing the "oadp-registry-config" which should have come with OADP installation.  See the log message below when I ran Backup:
-----------------------------------
...
time="2021-04-12T23:12:26Z" level=info msg="[common-backup] Error in getting route: configmaps \"oadp-registry-config\" not found. Assuming this is outside of OADP context." backup=oadp-operator/nginx-yc-backup2 cmd=/plugins/velero-plugins logSource="/go/src/github.com/konveyor/openshift-velero-plugin/velero-plugins/common/backup.go:48" pluginName=velero-plugins
...
-----------------------------------
[core@irv-12-181 ~]$ oc get configmap -n oadp-operator
NAME                           DATA   AGE
oadp-operator-lock             0      11d
restic-restore-action-config   1      11d

**Comment by shubham-pampattiwar** [comment]
@phuongatemc The `oadp-registry` as well as the `oadp-registry-config` are only created if you have a `BackupStorageLocation` (object storage) configured with your Velero instance. (`BackupStorageLocation` as AWS S3 is supported for registry creation by OADP). Also, this `oadp-registry` is only needed for backup of `imagestream` resources.

**Comment by phuongatemc** [comment]
@shubham-pampattiwar  Is there anyway we can create this oadp-registry-config separately from the creation/installation of Velero?  In our workflow, we install Velero first, and later, when user creates backup policy, we would create the BSL for that policy.
Yes, we need the oadp-registry when we do backup imagestream resources, that's one of our usage scenario.

**Comment by shubham-pampattiwar** [comment]
@phuongatemc In your workflow, the step at which you create the BSL, what I would suggest to so is update the Velero CR file with BSL details and then use the `oc apply -f <Velero_CR_File>` command on the file used for Velero CR creations. The OADP operator will take into consideration the updated Velero CR details having BSL configuration and in turn create the BSL, this will also trigger the creation of the `oadp-registry` as well as the `oadp-registry-config`. Does this workflow help your use case ? Let me know if this works out.

**Comment by phuongatemc** [comment]
@shubham-pampattiwar 
I tried to add "backup_storage_locations" but it seems like the oadp didn't pick up the BSL.  After I applied, oadp-registry-config or BSL was not created.  I think it mainly because of the BSL may have had something wrong in the config.  Here is the yaml I used. 

apiVersion: konveyor.openshift.io/v1alpha1
kind: Velero
metadata:
  name: example-velero
  namespace: oadp-operator
spec:
  olm_managed: false
  backup_storage_locations:
  - name: default
    accessMode: ReadWrite
    backupSyncPeriod: 0s
    config:
      ddHost: 10.235.110.197
      namespace: powerprotect
      prefix: PLCTLP-0602fb3e-3bf7-409a-8824-352a3722b50a/Stagings/fd21084d-5999-4b4f-a9cf-bd921efd3aa1/velerobackups
      secret: ph-test-plc-irv-vm-90-45-b8f80
    objectStorage:
      bucket: ph-test-plc-irv-vm-90-45-b8f80
      prefix: PLCTLP-0602fb3e-3bf7-409a-8824-352a3722b50a/Stagings/fd21084d-5999-4b4f-a9cf-bd921efd3aa1/velerobackups
    provider: ppdm.dellemc.com/velero-dd-plugin
    validationFrequency: 8760h0m0s
  custom_velero_plugins:
    - image: 'ap-sc.lss.emc.com:8446/dellemc/powerprotect-velero-dd:19.7.0-1-SNAPSHOT'
      name: powerprotect-velero-dd
    - image: ap-sc.lss.emc.com:8446/vsphereveleroplugin/velero-plugin-for-vsphere:1.1.0-rc4
      name: velero-plugin-for-vsphere
  default_velero_plugins:
    - openshift
    - csi
  enable_restic: true
  velero_feature_flags: EnableLocalMode

**Comment by shubham-pampattiwar** [comment]
@phuongatemc Just checked the BSL configuration you pasted. The provider for the BSL specified seems to be `ppdm.dellemc.com/velero-dd-plugin` . OADP currently only supports AWS, Azure and GCP as providers for BSL. Also, the creation of oadp-registry is supported only for `aws` as a provider for BSL. All in all, I would say if you configure your BSL using `aws` provider, it will work for your use case. 
Reference doc for BSL: https://github.com/konveyor/oadp-operator/blob/master/docs/bsl_and_vsl.md
Hope this helps.

**Comment by phuongatemc** [comment]
I was able to set up ASW S3 as BSL with OADP 0.2.0 as is (I will try 0.2.1 next), and ran a Backup on a test project.  It hits the error below when backing up ImageStream.  It looks like Velero cannot find the oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com which was set up by the OpenShift Route in the oadp-operator.
-------------------------------
time="2021-04-22T21:52:49Z" level=info msg="attempt #7 failed, waiting 35s and then retrying" backup=oadp-operator/test-velero2 cmd=/plugins/velero-plugins logSource="/go/src/github.com/konveyor/openshift-velero-plugin/vendor/github.com/bombsimon/logrusr/logrusr.go:100" pluginName=velero-plugins
time="2021-04-22T21:52:49Z" level=info msg="[imagecopy] Error copying image: Error trying to reuse blob sha256:469cfcc7a4b3947a4fa549c68cf4f8570be53779725f0c19f3d33d1520b08db0 at destination: error pinging docker registry oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com: Get \"http://oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com/v2/\": dial tcp: lookup oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com on 172.30.0.10:53: no such host" backup=oadp-operator/test-velero2 cmd=/plugins/velero-plugins logSource="/go/src/github.com/konveyor/openshift-velero-plugin/vendor/github.com/bombsimon/logrusr/logrusr.go:100" pluginName=velero-plugins
time="2021-04-22T21:52:49Z" level=info msg="1 errors encountered backup up item" backup=oadp-operator/test-velero2 logSource="pkg/backup/backup.go:451" name=ruby-hello-world
time="2021-04-22T21:52:49Z" level=error msg="Error backing up item" backup=oadp-operator/test-velero2 error="error executing custom action (groupResource=imagestreams.image.openshift.io, namespace=ph-hello, name=ruby-hello-world): rpc error: code = Unknown desc = Error trying to reuse blob sha256:469cfcc7a4b3947a4fa549c68cf4f8570be53779725f0c19f3d33d1520b08db0 at destination: error pinging docker registry oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com: Get \"http://oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com/v2/\": dial tcp: lookup oadp-default-aws-registry-route-oadp-operator.apps.ocpk8s1.ocpdell.com on 172.30.0.10:53: no such host" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/backup/item_backupper.go:331" error.function="github.com/vmware-tanzu/velero/pkg/backup.(*itemBackupper).executeActions" logSource="pkg/backup/backup.go:455" name=ruby-hello-world
